src/Orleans.Transactions/DistributedTM/Implementation/TransactionParticipant.cs

```diff
@@ -0,0 +1,904 @@
+                                lockWorker.Notify();
+                            }
+                            else
+                            {
+                                storageWorker.Notify(bottom.WaitingSince + PrepareTimeout);
+                                if (logger.IsEnabled(LogLevel.Trace))
+                                    logger.Trace($"{commitQueue.Count} waiting on: LocalCommitEntry {bottom.Timestamp:o} WaitCount={bottom.WaitCount}");
+                            }
+                            break;
+                        }
+
+                    case CommitRole.RemoteCommit:
+                        {
+                            if (bottom.PrepareIsPersisted && !bottom.LastSent.HasValue)
+                            {
+                                // send PreparedMessage to remote TM
+                                bottom.TransactionManager.Prepared(bottom.TransactionId, bottom.Timestamp, thisParticipant, TransactionalStatus.Ok).Ignore();                                
+                                    
+                                bottom.LastSent = now;
+
+                                if (bottom.IsReadOnly)
+                                {
+                                    storageWorker.Notify(); // we are ready to batch now
+                                }
+                                else
+                                {
+                                    storageWorker.Notify(bottom.LastSent.Value + RemoteTransactionPingFrequency);
+                                }
+                            }
+                            else if (!bottom.IsReadOnly && bottom.LastSent.HasValue)
+                            {
+                                // send ping messages periodically to reactivate crashed TMs
+
+                                if (bottom.LastSent + RemoteTransactionPingFrequency <= now)
+                                {
+                                    bottom.TransactionManager.Ping(bottom.TransactionId, bottom.Timestamp, thisParticipant);
+                                    bottom.LastSent = now;
+                                }
+                                else
+                                {
+                                    storageWorker.Notify(bottom.LastSent.Value + RemoteTransactionPingFrequency);
+                                }
+                            }
+
+                            if (logger.IsEnabled(LogLevel.Trace))
+                                logger.Trace($"{commitQueue.Count} waiting on: RemoteCommitEntry {bottom.Timestamp:o} IsReadOnly={bottom.IsReadOnly} PrepareIsPersisted={bottom.PrepareIsPersisted} LastSent={bottom.LastSent}");
+
+                            break;
+                        }
+
+                    default:
+                        {
+                            logger.Error(777, $"internal error: impossible case");
+                            throw new NotImplementedException();
```

Comment:
suggest throwing not supported exception with bottom.Role value used in message body?  I understand this should not happen, but if it does it would be useful to know which case was not handled.  May also be valuable in the log line.